File: leetcode/p384.py
class Solution:

    def __init__(self, nums: List[int]):
        self.shuffled = nums[::]
        self.nums = nums[::]

    def reset(self) -> List[int]:
        """
        Resets the array to its original configuration and return it.
        """
        return self.nums

    def shuffle(self) -> List[int]:
        """
        Returns a random shuffling of the array.
        """
        for i in range(len(self.shuffled)):
            n = random.randint(0, i)
            t = self.shuffled[n]
            self.shuffled[n] = self.shuffled[i]
            self.shuffled[i] = t
        return self.shuffled

# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()

File: leetcode/p384.cpp
// https://leetcode.com/problems/shuffle-an-array/
// 384. Shuffle an Array

class Solution {
public:
    vector<int>& array;
    Solution(vector<int>& nums):array(nums) {
        return;
    }
    
    /** Resets the array to its original configuration and return it. */
    vector<int> reset() {
        return array;
    }
    
    /** Returns a random shuffling of the array. */
    vector<int> shuffle() {
        vector<int> res(array);
        int n = res.size();
        for (int i=0;i<n;i++) {
            int t = rand()%(n-i) + i;
            swap(res[i], res[t]);
        }
        return res;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */
